### PR TITLE
Fix/tao 9835/fix replace result namespace

### PR DESCRIPTION
**Pull request by siwane**
In order to follow ltiv1p1 spec, replaceResult xml must be adapted to follow specifications.

LTI specificications: https://www.imsglobal.org/specs/ltiv1p1p1/implementation-guide#toc-26
Ticket: https://oat-sa.atlassian.net/browse/TAO-9835

### CODE REVIEW

File: test/Unit/helpers/LtiBasicOutcomeTest.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            $grade,
+            $sourcedIdNode = $xpath->query('//ns:textString')->item(0)->nodeValue
+        );
+
+        $this->assertEquals(
+            'en-us',
+            $sourcedIdNode = $xpath->query('//ns:language')->item(0)->nodeValue
+        );
+    }
+}
```
Comment (mike-ionut-mihai-sandu-tao) [inline test/Unit/helpers/LtiBasicOutcomeTest.php:68]: \n

File: manifest.php [hunk 2]
```diff
@@ -1,30 +1,45 @@
...
 
 return [
     'name' => 'taoLtiBasicOutcome',
-  'label' => 'Result storage for LTI',
+    'label' => 'Result storage for LTI',
     'description' => 'Implements the LTI basic outcome engine for LTI Result Server',
-  'license' => 'GPL-2.0',
-  'version' => '3.4.1',
+    'license' => 'GPL-2.0',
+    'version' => '3.4.3',
```
Comment (mike-ionut-mihai-sandu-tao) [inline manifest.php:29]: why from '3.4.1', to '3.4.3', is '3.4.2', used?

File: test/Unit/helpers/LtiBasicOutcomeTest.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+use DOMDocument;
+use DOMXpath;
+use oat\generis\test\TestCase;
+use taoLtiBasicOutcome_helpers_LtiBasicOutcome as LtiBasicOutcome;
+
+class LtiBasicOutcomeTest extends TestCase
+{
+    public function testBuildXMLMessage(): void
+    {
+        $sourcedId = 'toto';
```
Comment (mike-ionut-mihai-sandu-tao) [inline test/Unit/helpers/LtiBasicOutcomeTest.php:35]: :)

File: test/Unit/helpers/LtiBasicOutcomeTest.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+declare(strict_types=1);
+
+namespace oat\ltiBasicOutcome\test\Unit\helpers;
+
+use DOMDocument;
+use DOMXpath;
+use oat\generis\test\TestCase;
+use taoLtiBasicOutcome_helpers_LtiBasicOutcome as LtiBasicOutcome;
+
+class LtiBasicOutcomeTest extends TestCase
```
Comment (mike-ionut-mihai-sandu-tao) [inline test/Unit/helpers/LtiBasicOutcomeTest.php:31]: please run phpcs on this

Comment (siwane) [reply test/Unit/helpers/LtiBasicOutcomeTest.php:31]: done

File: manifest.php (hunk 2 above)
Comment (siwane) [reply manifest.php]: good catch

File: test/Unit/helpers/LtiBasicOutcomeTest.php (hunk 1 above)
Comment (siwane) [reply test/Unit/helpers/LtiBasicOutcomeTest.php]: done

**Review by gabrielfs7 — COMMENTED** [review]
Just a suggestion to build this XML body in a central place, cause now we have to fix 3 places.

- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

File: manifest.php [hunk 5]
```diff
@@ -1,30 +1,45 @@
...
 
 return [
     'name' => 'taoLtiBasicOutcome',
-  'label' => 'Result storage for LTI',
+    'label' => 'Result storage for LTI',
     'description' => 'Implements the LTI basic outcome engine for LTI Result Server',
-  'license' => 'GPL-2.0',
-  'version' => '3.4.1',
+    'license' => 'GPL-2.0',
+    'version' => '3.4.2',
```
Comment (gabrielfs7) [inline manifest.php:29]: Isn it a breaking change now that File was removed? One way it to keep the file and do a delegate for the new factory inside. Wdyt?

File: models/classes/class.LtiBasicOutcome.php [hunk 6]
```diff
@@ -69,8 +70,13 @@ public function storeTestVariable($deliveryResultIdentifier, $test, taoResultSer
...
-                $message = taoLtiBasicOutcome_helpers_LtiBasicOutcome::buildXMLMessage($deliveryResultIdentifier, $grade, 'replaceResultRequest');
-
+                $ltiOutcomeXmlFactory = $this->getLtiOutcomeXmlFactory();
+                $message = $ltiOutcomeXmlFactory->build(
+                    $deliveryResultIdentifier,
+                    $grade,
+                    uniqid('', true),
+                    'replaceResultRequest'
+                );
                 //common_Logger::i("Preparing POX message for the outcome service :".$message."\n");
```
Comment (gabrielfs7) [inline models/classes/class.LtiBasicOutcome.php:80]: Not related to your PR, but I think we can remove this comment.

File: models/classes/class.LtiBasicOutcome.php [hunk 7]
```diff
@@ -69,8 +70,13 @@ public function storeTestVariable($deliveryResultIdentifier, $test, taoResultSer
                 $deliveryResultAlias = $resultAliasService->getResultAlias($deliveryResultIdentifier);
                 $deliveryResultIdentifier = empty($deliveryResultAlias) ? $deliveryResultIdentifier : current($deliveryResultAlias);
 
-                $message = taoLtiBasicOutcome_helpers_LtiBasicOutcome::buildXMLMessage($deliveryResultIdentifier, $grade, 'replaceResultRequest');
-
+                $ltiOutcomeXmlFactory = $this->getLtiOutcomeXmlFactory();
```
Comment (gabrielfs7) [inline models/classes/class.LtiBasicOutcome.php:73]: Thanks 🙇

File: models/classes/class.LtiBasicOutcome.php [hunk 8]
```diff
@@ -69,14 +70,16 @@ public function storeTestVariable($deliveryResultIdentifier, $test, taoResultSer
...
 
-                $message = taoLtiBasicOutcome_helpers_LtiBasicOutcome::buildXMLMessage($deliveryResultIdentifier, $grade, 'replaceResultRequest');
-
-                //common_Logger::i("Preparing POX message for the outcome service :".$message."\n");
+                $ltiOutcomeXmlFactory = $this->getLtiOutcomeXmlFactory();
+                $message = $ltiOutcomeXmlFactory->buildReplaceResultRequest(
+                    $deliveryResultIdentifier,
+                    $grade,
+                    uniqid('', true),
+                    'replaceResultRequest'
```
Comment (gabrielfs7) [inline models/classes/class.LtiBasicOutcome.php:78]: This should be language parameter or omitted, no?

**Review by gabrielfs7 — APPROVED** [review]
LGTM, just one parameter order it wrong as I mentioned in the comment.

File: models/classes/class.LtiBasicOutcome.php (hunk 8 above)
Comment (mike-ionut-mihai-sandu-tao) [reply models/classes/class.LtiBasicOutcome.php]: fixed

File: models/classes/class.LtiBasicOutcome.php [hunk 9]
```diff
@@ -156,4 +158,9 @@ public function storeItemVariables($deliveryResultIdentifier, $test, $item, arra
             $this->storeItemVariable($deliveryResultIdentifier, $test, $item, $itemVariable, $callIdItem);
         }
     }
+
+    private function getLtiOutcomeXmlFactory(): LtiOutcomeXmlFactory
+    {
+        $this->getServiceLocator()->get(LtiOutcomeXmlFactory::class);
```
Comment (boajer) [inline models/classes/class.LtiBasicOutcome.php:164]: I think that is the must here
```suggestion
       return $this->getServiceLocator()->get(LtiOutcomeXmlFactory::class);
```

**Review by boajer — CHANGES_REQUESTED** [review]
A minor change required

File: manifest.php [hunk 10]
```diff
@@ -1,30 +1,45 @@
...
     'description' => 'Implements the LTI basic outcome engine for LTI Result Server',
-  'license' => 'GPL-2.0',
-  'version' => '3.4.1',
+    'license' => 'GPL-2.0',
+    'version' => '4.0.0',
     'author' => 'Open Assessment Technologies',
     'requires' => [
         'taoResultServer' => '>=4.2.0',
-        'taoLti' => '>=5.0.0'
+        'taoLti' => '>=5.0.0',
```
Comment (boajer) [inline manifest.php:33]:
```suggestion
        'taoLti' => '>=11.12.0',
```
due to introduction of `LtiOutcomeXmlFactory`

File: models/classes/class.LtiBasicOutcome.php [hunk 11]
```diff
@@ -69,14 +70,15 @@ public function storeTestVariable($deliveryResultIdentifier, $test, taoResultSer
...
                 $deliveryResultIdentifier = empty($deliveryResultAlias) ? $deliveryResultIdentifier : current($deliveryResultAlias);
 
-                $message = taoLtiBasicOutcome_helpers_LtiBasicOutcome::buildXMLMessage($deliveryResultIdentifier, $grade, 'replaceResultRequest');
-
-                //common_Logger::i("Preparing POX message for the outcome service :".$message."\n");
+                $ltiOutcomeXmlFactory = $this->getLtiOutcomeXmlFactory();
+                $message = $ltiOutcomeXmlFactory->buildReplaceResultRequest(
+                    $deliveryResultIdentifier,
+                    $grade,
+                    uniqid('', true)
```
Comment (boajer) [inline models/classes/class.LtiBasicOutcome.php:77]: If we are going towards random, I would prefer `Uuid::uuid4()`

**Review by boajer — CHANGES_REQUESTED** [review]
1 minor change needed + 1 recommendation for discussion :)

File: models/classes/class.LtiBasicOutcome.php (hunk 11 above)
Comment (mike-ionut-mihai-sandu-tao) [reply models/classes/class.LtiBasicOutcome.php:77]: it was the previous way of computing the id , but let me check if i can implement effortless Uuid::uuid4()

**Review by boajer — APPROVED** [review]
LGFM

**Pull request merged**